### PR TITLE
Add recursive dataset mounting support to pam_zfs_key

### PR DESCRIPTION
**Pull request by jkolo**
### PR Title
Add recursive dataset mounting support to pam_zfs_key
### Motivation and Context
This change introduces the ability to recursively mount ZFS datasets using a new configuration option, `mount_recursively`. This enhancement solves the need for hierarchical or nested dataset management in environments that utilize `pam_zfs_key`, ensuring all eligible child datasets are mounted as part of the session functionality.
This feature improves the flexibility and usability of `pam_zfs_key` when managing complex dataset organizations.
### Description
- Added support for a new configuration option, `mount_recursively`, which ensures child datasets of a given root can also be mounted.
- Refactored the `decrypt_mount()` function to integrate recursive mounting functionality, leveraging `zfs_iter_filesystems_v2` for child dataset traversal.
- Updated the dataset mounting logic to ensure properties such as `canmount` and `mountpoint` are respected during recursive mounting.
- Modified configuration handling (via `zfs_key_config_t`) to include the `mount_recursively` option.
- Enhanced the session logic to include recursive mounting without breaking existing behavior.
- Added a functional test, `pam_mount_recursively.ksh`, to validate recursive mounting behavior and edge cases (e.g., "none", "legacy", "noauto", and "off" mountpoints).

### How Has This Been Tested?
- Added a functional test (`tests/zfs-tests/tests/functional/pam/pam_mount_recursively.ksh`) that:
    - Validates recursive mounting for nested datasets with the `mount_recursively` flag enabled.
    - Tests behavior across various child dataset configurations (e.g., different `canmount` and `mountpoint` properties).
    - Ensures correct key loading status through session management (`open_session` and `close_session`) via PAM.
    - Verifies datasets are only mounted when they satisfy `canmount=on` and appropriate mountpoint properties.

- Manual testing performed on Linux configurations using `pamtester` to ensure real-world applicability.
- Verified backward compatibility where `mount_recursively` is not enabled.

### Types of Changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly. _(Implementation details documented in code comments. Additional documentation updates suggested separately if needed.)_
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

Let me know if additional details are required.

### CODE REVIEW

**Comment by jkolo** [comment]
Pull request is ready for merge. Is there anything else I should do to get it to pass review? @behlendorf dorry for calling directly. I didn't know who or where best to ask :)

File: contrib/pam_zfs_key/pam_zfs_key.c [hunk 1]
```diff
@@ -508,6 +452,217 @@ zfs_key_config_load(pam_handle_t *pamh, zfs_key_config_t *config,
...
+		    zfs_get_name(zhp));
+		return (0);
+	}
+
+	/* Check if prop canmount is on */
+	if (zfs_prop_get_int(zhp, ZFS_PROP_CANMOUNT) != ZFS_CANMOUNT_ON) {
+		pam_syslog(pamh, LOG_INFO,
+		    "canmount is not on for: %s, skipping",
+		    zfs_get_name(zhp));
+		return (0); // Skip dataset
```
Comment (behlendorf) [inline contrib/pam_zfs_key/pam_zfs_key.c:495]:
```suggestion
		return (0);
```

To be consistent with the other return 0's.  Unless this one is special in some way?

File: contrib/pam_zfs_key/pam_zfs_key.c [hunk 2]
```diff
@@ -508,6 +452,217 @@ zfs_key_config_load(pam_handle_t *pamh, zfs_key_config_t *config,
...
+		return (-1);
+	}
+
+	/* Check if mountpoint isn't none or legacy */
+	if (strcmp(mountpoint, ZFS_MOUNTPOINT_NONE) == 0 ||
+	    strcmp(mountpoint, ZFS_MOUNTPOINT_LEGACY) == 0) {
+		pam_syslog(pamh, LOG_INFO,
+		    "mountpoint is none or legacy for: %s, skipping",
+		    zfs_get_name(zhp));
+		return (0);  // Skip dataset
```
Comment (behlendorf) [inline contrib/pam_zfs_key/pam_zfs_key.c:513]:
```suggestion
		return (0);
```

**Review by behlendorf — APPROVED** [review]
This makes sense.  Thanks for working on this and adding the test case.

File: contrib/pam_zfs_key/pam_zfs_key.c (hunk 1 above)
Comment (jkolo) [reply contrib/pam_zfs_key/pam_zfs_key.c:495]: No, it is not unique. Corrected.

File: contrib/pam_zfs_key/pam_zfs_key.c (hunk 2 above)
Comment (jkolo) [reply contrib/pam_zfs_key/pam_zfs_key.c:513]: No, it is not unique. Corrected.

**Comment by jkolo** [comment]
@behlendorf friendly reminder

File: tests/zfs-tests/tests/functional/pam/pam_mount_recursively.ksh [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	log_must pamtester ${pamservice} ${username} close_session
+	references 0
+	log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}"
+    log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}/deep"
+    log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}/deep/deeper"
+    log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}/deep/none"
+    log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}/deep/noauto"
+    log_mustnot ismounted "$TESTPOOL/mrec/pam/${username}/deep/off"
+	keystatus unavailable
+}
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/pam/pam_mount_recursively.ksh:73]: Check indent

**Comment by tonyhutter** [comment]
@jkolo sorry this PR fell though the cracks.  This should be ready to go once the indents are fixed.

File: tests/zfs-tests/tests/functional/pam/pam_mount_recursively.ksh (hunk 3 above)
Comment (jkolo) [reply tests/zfs-tests/tests/functional/pam/pam_mount_recursively.ksh]: Fixed

**Comment by jkolo** [comment]
@tonyhutter  and @behlendorf fixed and rebased

**Review by tonyhutter — APPROVED** [review]
Thanks, can you squash the commits too?

**Comment by jkolo** [comment]
> Thanks, can you squash the commits too?

@tonyhutter commits have been squashed
